api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -276,4 +359,8 @@ message Resource {
   // Cache control properties for the resource.
   // [#not-implemented-hide:]
   CacheControl cache_control = 7;
+
+  // Dynamic parameter constraints associated with this resource. To be used by client-side caches
+  // (including xDS proxies) when matching subscribed resource locators.
+  repeated DynamicParameterConstraints dynamic_parameter_constraints = 8;
```

Comment:
IIUC in a resource response all constraints will be sent back to the client, even the ones it didn’t ask for (e.g., client asks for a resource with `{foo=true}`, and gets back the resource R with `{foo=true, bar=false, minor_version={min=4, max=5}, …}`.

A couple of questions:
1. Is this the intended behavior? (IMHO yes, because an xdd-relay may need these constraints for future clients)
2. Do we have an estimate on the impact on response sizes, or an estimate on the number of constraints that a resource may have?

If all the constraints are only needed for xds-relays, and not for end-clients, then I think only they should pay that cost, and an end-client can just receive the relevant constraints (or no constraints, if it assumes the server returns a valid response which doesn’t break the constraints).

Reply:
I'm still working on nailing down the details of the matching semantics for cases where keys are missing, so the exact example you cite might not actually wind up being possible.

But to answer your more general question, yes, it is the intended behavior to send the constraints to the client, specifically because it is needed by caching xDS proxies.

I don't have any data on how much overhead there will be here.  If it becomes a problem, we could probably introduce a client feature to indicate that the client is not a caching proxy and therefore the server can leave this field unset.